repair/repair.cc

```diff
@@ -753,32 +757,36 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
         // repair all the ranges in limited parallelism
         rlogger.info("repair[{}]: Started to repair {} out of {} tables in keyspace={}, table={}, table_id={}, repair_reason={}",
                 ri->id.uuid, idx + 1, ri->table_ids.size(), ri->keyspace, table_name, table_id, ri->reason);
-        co_await coroutine::parallel_for_each(ri->ranges, [ri, table_id] (auto&& range) {
-            return with_semaphore(ri->rs.repair_module()->range_parallelism_semaphore(), 1, [ri, &range, table_id] {
-                return ri->repair_range(range, table_id).then([ri] {
-                    if (ri->reason == streaming::stream_reason::bootstrap) {
-                        ri->rs.get_metrics().bootstrap_finished_ranges++;
-                    } else if (ri->reason == streaming::stream_reason::replace) {
-                        ri->rs.get_metrics().replace_finished_ranges++;
-                    } else if (ri->reason == streaming::stream_reason::rebuild) {
-                        ri->rs.get_metrics().rebuild_finished_ranges++;
-                    } else if (ri->reason == streaming::stream_reason::decommission) {
-                        ri->rs.get_metrics().decommission_finished_ranges++;
-                    } else if (ri->reason == streaming::stream_reason::removenode) {
-                        ri->rs.get_metrics().removenode_finished_ranges++;
-                    } else if (ri->reason == streaming::stream_reason::repair) {
-                        ri->rs.get_metrics().repair_finished_ranges_sum++;
-                        ri->nr_ranges_finished++;
-                    }
-                    rlogger.debug("repair[{}]: node ops progress bootstrap={}, replace={}, rebuild={}, decommission={}, removenode={}, repair={}",
-                        ri->id.uuid,
-                        ri->rs.get_metrics().bootstrap_finished_percentage(),
-                        ri->rs.get_metrics().replace_finished_percentage(),
-                        ri->rs.get_metrics().rebuild_finished_percentage(),
-                        ri->rs.get_metrics().decommission_finished_percentage(),
-                        ri->rs.get_metrics().removenode_finished_percentage(),
-                        ri->rs.get_metrics().repair_finished_percentage());
-                });
+        co_await coroutine::parallel_for_each(ri->ranges, [tmp_ri = ri, table_id, task_id] (auto&& range) -> future<> {
+            auto ri = tmp_ri;
+            auto task_impl_ptr = std::make_unique<tasks::range_repair_task_impl>(ri->rs.repair_module(), tasks::task_id::create_null_id(), 0, ri->keyspace, "", format("{}", ri->reason), "", task_id, ri, range, table_id);
+            auto id = co_await ri->rs.repair_module()->make_task(this_shard_id(), std::move(task_impl_ptr), tasks::task_manager::parent_data{task_id, this_shard_id()});
+            auto task = ri->rs.repair_module()->get_tasks()[id];
```

Comment:
I wonder if make_task should return the id or the task foreign shared ptr then
you did the same for other tasks as well and if this is the common use case maybe we should change `make_task` to return what its callers need.

Reply:
there should be no problem accessing the task->id() on the same shard and even on a foreign shard.

Reply:
I think we should make `make_task` return `task_ptr` (and not `foreign_task_ptr`), since it gets a pointer to task impl as an argument and we should rather allow that only on target shard than play with foreign's.

The old `make_task` should stay as it is though I believe.

Reply:
makes sense